portfolio/src/main/java/com/google/sps/data/HighestPriorityFirstScheduler.java

```diff
@@ -0,0 +1,124 @@
+
+    List<ScheduledTask> scheduledTasks = new ArrayList<ScheduledTask>();
+
+    int availableTimesIndex = 0;
+
+    Instant currentScheduleTime = workHoursStartTime;
+
+    while (availableTimesIndex < currentAvailableTimes.size() && !taskQueue.isEmpty()) {
+      TimeRange availableTimeRange = currentAvailableTimes.get(availableTimesIndex);
+      Task task = taskQueue.peek();
+
+      if (availableTimeRange.start().isAfter(currentScheduleTime)) {
+        currentScheduleTime = availableTimeRange.start();
+      }
+
+      Boolean doesTaskFit =
+          !currentScheduleTime
+              .plusSeconds(task.getDuration().getSeconds())
+              .isAfter(availableTimeRange.end());
+
+      if (doesTaskFit) {
+        ScheduledTask scheduledTask =
+            new ScheduledTask(
+                task, currentScheduleTime, /* isCompletelyScheduled = */ Optional.of(true));
+        scheduledTasks.add(scheduledTask);
+
+        // Delete the TimeRange that has been scheduled over so that different priority tasks
+        // won't be scheduled over the same time.
+        availableTimesGroup.deleteTimeRange(
+            TimeRange.fromStartEnd(
+                scheduledTask.getStartTime(),
+                scheduledTask
+                    .getStartTime()
+                    .plusSeconds(scheduledTask.getTask().getDuration().getSeconds())));
+        // After deleting a TimeRange, we must renew the list.
+        currentAvailableTimes = getAvailableTimeRangesList(availableTimesGroup);
+
+        // Push the currentScheduleTime back by the amount that was scheduled.
+        currentScheduleTime = currentScheduleTime.plusSeconds(task.getDuration().getSeconds());
+        taskQueue.remove();
+
+        // If the next task's priority is different from the task that was just scheduled,
+        // then reset the availableTimesIndex, currentScheduleTime, and make the loop
+        // grab a TimeRange. We go back to the first availableTimeRange in order to schedule
+        // more tasks towards the beginning of the availableTimes.
+        if (isNextTaskDifferentPriority(taskQueue, task)) {
+          availableTimesIndex = 0;
+          currentScheduleTime = workHoursStartTime;
+        }
+        // If the task's priority is not different, then we can simply continue running the
+        // the loop without retrieving the next availableTimeRange since there might be
+        // some time left over towards the end of the current one.
+      } else if (availableTimesIndex == currentAvailableTimes.size() - 1) {
+        // If we've reached the end of the availableTimeRange group and we can't schedule a task,
+        // then we can remove all the remaining tasks of equal priority since they will all be
+        // longer in duration therefore, they will not be able to be scheduled either.
+        if (!taskQueue.isEmpty()) {
+          availableTimesIndex = 0;
+          while (!taskQueue.isEmpty()
```

Comment:
Consider using a helper method for this while loop.

ex:
private void removeTasksWithPriority(TaskQueue taskQueue, TaskPriority taskPriority) {
  ...
}

Reply:
Done.